include/net/buf.h

```diff
@@ -455,6 +455,54 @@ static inline void net_buf_simple_restore(struct net_buf_simple *buf,
  */
 #define NET_BUF_EXTERNAL_DATA  BIT(1)
 
+/** Flag indicating that both the buffer and its associated data pointer,
+ * point to externally allocated memory and do not come from any pool.
+ * Therefore once ref goes down to zero, the pointed  buffer and data
+ * will not need to be deallocated. This never needs to be
+ * explicitly set or unset by the net_buf API user. Such net_buf is
+ * exclusively instantiated via net_buf_standalone_setup() function..
+ */
+#define NET_BUF_STANDALONE BIT(2)
+
+/** @def NET_BUF_DEFINE
+ *  @brief Define a net_buf stack variable.
+ *
+ *  This is a helper macro which is used to define a net_buf object
+ *  on the stack.
+ *
+ *  @param _name Name of the net_buf object.
+ *  @param _size Maximum data storage for the buffer.
+ */
+#define NET_BUF_DEFINE(_name, _size)				\
```

Comment:
Are you really seeing the need for this? I'd imagine only the static variant to be useful, since if you need something only for the duration for the function's lifetime (that the variable is declared in) then net_buf_simple should be the way to go.

Reply:
Also, I'm a bit uncomfortable with how standard/normal the name of this looks like, considering that this is a bit of a corner case that will only have a few users. Meaning, it might be better to have "STANDALONE" in the API name, to make it clear that this is something "special" and not something that most users of net_buf will need.

Reply:
@jhedberg 

> then net_buf_simple should be the way to go.

net_buf_simple is NOT used anywhere in the IP stack code. As it stands currently, net_buf_simple can as well be renamed to net_buf_adhoc_variant_for_bt_subsys ;-).

Reply:
I'm questioning here the introduction of an on-stack net_buf variable. I don't think there's so far been any use cases for that in the network subsystem. With such a variable you want to be sure that no-one has grabbed references to it by the time you return from your function (in whose scope the variable was declared in). net_buf_simple gives that, since there are no APIs (like net_buf_ref) that would give other functions the idea that they can "own" or hold onto the buffer.

As for the static net_buf variable, that's something I'm less concerned of, but I still fail to see how we can avoid buggy code if there's no way to catch the point when the buffer can be reused (and if the code has to wait for this, then I don't see a different with having a dedicated one-buffer net_buf_pool).

Btw, any net_buf using code does use net_buf_simple since that's how the data is internally stored in net_buf. This is particularly convenient when you have a net_buf and want to pass the data to some sub-function, but not give it the opportunity to grab a reference or mess with other net_buf state. In such a case you just pass `&buf->b`, i.e. a pointer to the net_buf_simple.

Reply:
> With such a variable you want to be sure that no-one has grabbed references to it by the time you return from your function

For reference, my understanding how this is handled is that refcounted buffers are counter-incremented, and of non-refcounted, a copy is made.

> I'm questioning here the introduction of an on-stack net_buf variable.

My understanding of this patch is that @tbursztyka wants to introduce a complete interface, to be available to any client whenever (if) needed (all prompted by his larger WIP refactor of buf handling in the IP code).

> Btw, any net_buf using code does use net_buf_simple since

Cool, and yet ident "net_buf_simple" is never referenced in subsys/net/ip/.


(All these responses are my trying to understand these (fairly complex, if not confusing) matters in grey-box manner. I definitely don't argue some point. (Except that all stuff should be understandable by other parties of course ;-) )).

Reply:
> Cool, and yet ident "net_buf_simple" is never referenced in subsys/net/ip/.

There has been so far no need to use the net_buf_simple API in IP stack side of the code.